spatialdata/_core/transform.py

```diff
@@ -1,86 +1,526 @@
-from functools import singledispatch
-from typing import Any, Optional
+from __future__ import annotations
+
+import copy
+import json
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List, Optional, Type, Union
 
-import anndata as ad
 import numpy as np
-import xarray as xr
 
 from spatialdata._types import ArrayLike
 
+__all__ = [
+    "BaseTransformation",
+    "Identity",
+    "MapIndex",
+    "MapAxis",
+    "Translation",
+    "Scale",
+    "Affine",
+    "Rotation",
+    "Sequence",
+    "Displacements",
+    "Coordinates",
+    "VectorField",
+    "InverseOf",
+    "Bijection",
+    "ByDimension",
+    "compose_transformations",
+]
+
+
+class BaseTransformation(ABC):
+    @property
+    @abstractmethod
+    def ndim(self) -> Optional[int]:
+        pass
+
+    @abstractmethod
+    def to_dict(self) -> Dict[str, Any]:
+        pass
+
+    def to_json(self) -> str:
+        return json.dumps(self.to_dict())
+
+    def compose_with(self, transformation: BaseTransformation) -> BaseTransformation:
+        return compose_transformations(self, transformation)
+
+    @abstractmethod
+    def transform_points(self, points: ArrayLike) -> ArrayLike:
+        pass
+
+    @abstractmethod
+    def inverse(self) -> BaseTransformation:
+        pass
+
+
+def compose_transformations(transformation0: BaseTransformation, transformation1: BaseTransformation) -> Sequence:
+    return Sequence([transformation0, transformation1])
+
+
+def get_transformation_from_json(s: str) -> BaseTransformation:
+    d = json.loads(s)
+    return get_transformation_from_dict(d)
+
+
+def get_transformation_from_dict(d: Dict[str, Any]) -> BaseTransformation:
+    kw = d.copy()
+    type = kw["type"]
+    cls: Type[BaseTransformation]
+    if type == "identity":
+        cls = Identity
+    elif type == "mapIndex":
+        cls = MapIndex  # type: ignore
+    elif type == "mapAxis":
+        cls = MapAxis  # type: ignore
+    elif type == "translation":
+        cls = Translation
+    elif type == "scale":
+        cls = Scale
+    elif type == "affine":
+        cls = Affine
+    elif type == "rotation":
+        cls = Rotation
+    elif type == "sequence":
+        cls = Sequence
+    elif type == "displacements":
+        cls = Displacements  # type: ignore
+    elif type == "coordinates":
+        cls = Coordinates  # type: ignore
+    elif type == "vectorField":
+        cls = VectorField  # type: ignore
+    elif type == "inverseOf":
+        cls = InverseOf
+    elif type == "bijection":
+        cls = Bijection
+    elif type == "byDimension":
+        cls = ByDimension  # type: ignore
+    else:
+        raise ValueError(f"Unknown transformation type: {type}")
+    del kw["type"]
+    if "input" in kw:
+        del kw["input"]
+    if "output" in kw:
+        del kw["output"]
```

Comment:
what is this? why `del` ?

Reply:
from the transform specs, there are cases where `input` and `output` (the input and output coordinate systems of the transformation) are specified in the dict describing the transformation. But in one edge case they are not specified: when a transformation is used within a sequence. In this file I am trying to separate the concept of transformation and of input/output coordinate system (as it is done in the specs), that's why if input and output are present we remove them, since the constructors of the various classes don't take them.

Reply:
We can safely use `del` since `kw` is a copy of the input dict.